pkg/credentialprovider/apis/criauth/v1alpha1/types.go

```diff
@@ -0,0 +1,148 @@
+type PluginArg struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginRequest includes the image that the kubelet requires authenication for.
+// Kubelet will pass this request object to the plugin via stdin.
+type CRIAuthPluginRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// image is the container image that is being pulled as part of the
+	// CRI auth plugin request. Plugins may optionally parse the image
+	// to extract any information required to fetch credentials.
+	Image string `json:"image"`
+}
+
+type PluginCacheKeyType string
+
+const (
+	// ImagePluginCacheKeyType means the kubelet will cache credentials on a per-image basis,
+	// using the image passed from the kubelet directly as the cache key. This includes
+	// the registry domain, port (if specified), and path but does not include tags or SHAs.
+	ImagePluginCacheKeyType PluginCacheKeyType = "Image"
+	// RegistryPluginCacheKeyType means the kubelet will cache credentials on a per-registry basis.
+	// The cache key will be based on the registry domain and port (if present) parsed from the requested image.
+	RegistryPluginCacheKeyType PluginCacheKeyType = "Registry"
+	// GlobalPluginCacheKeyType means the kubelet will cache credentials for all images that
+	// match for a given plugin. This cache key should only be returned by plugins that do not use
+	// the image input at all.
+	GlobalPluginCacheKeyType PluginCacheKeyType = "Global"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginResponse holds credentials that the kubelet should use for the specified
+// image provided in the original request. Kubelet will read the response from the plugin via stdout.
+type CRIAuthPluginResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// cacheKeyType indiciates the type of caching key to use based on the image provided
+	// in the request. There are three valid values for the cache key type: Image, Registry, and
+	// Global.
+	CacheKeyType PluginCacheKeyType `json:"cacheKeyType"`
+
+	// cacheDuration indicates the duration the provided credentials should be cached for.
+	// The kubelet will use this field to set the in-memory cache duration for credentials
+	// in the AuthConfig. If nil, the kubelet will use defaultCacheDuration provided in
+	// CRIAuthPluginConfig. If set to 0, the kubelet will not cache the provided AuthConfig.
+	CacheDuration *metav1.Duration `json:"cacheDuration,omitempty"`
+
+	// authConfig contains authentication information for the container registry.
+	// Only username/password based authentication is supported today, but more authentication
+	// mechanisms may be added in the future. A plugin should return a nil AuthConfig if
+	// no valid credentials can be returned for the requested image.
+	AuthConfig *AuthConfig `json:"authConfig,omitempty"`
```

Comment:
@liggitt in a previous discussion we had, we agreed that this should be `[]AuthConfig` so that implementations like azure can return multiple credentials for a given match. As I was implementing this I realized this won't work because the output of the plugin `credentialprovider.DockerConfig` is a map so a single match key can't map to multiple AuthConfigs:

https://github.com/kubernetes/kubernetes/blob/c82d5ee0486191e709663b584b4f7101235514c9/pkg/credentialprovider/config.go#L47-L50

I tried to just modify `DockerConfig` to something like `type DockerConfig map[string][]DockerConfigEntry` but this would break json serialization of the docker config and would require a larger re-work of all the existing provider implementations. The only way to map 1 match image to N AuthConfig is if the match comes from a separately registered provider:

https://github.com/kubernetes/kubernetes/blob/c82d5ee0486191e709663b584b4f7101235514c9/pkg/credentialprovider/keyring.go#L39

So I've left this to be just  `*AuthConfig` which means if an implementation needs to return multiple credentials for a given registry or match, they would have to register two separate plugins. For Azure specifically, it would look something like:
```
kind: CRIAuthPluginConfig
apiVersion: criauth.k8s.io/v1alpha1
plugins:
  - name: acr
    matchImages:
    - "*.azurecr.io"
    - "*.azurecr.cn"
    - "*.azurecr.de"
    - "*.azurecr.us"
    defaultCacheDuration: 1m
    extraArgs:
      azure-container-registry-config: "/etc/azure/azure-registry-config.yaml"
  - name: acr-anonymous
    matchImages:
    - "*.azurecr.io"
    - "*.azurecr.cn"
    - "*.azurecr.de"
    - "*.azurecr.us"
    defaultCacheDuration: 1m
    extraArgs:
      azure-container-registry-config: "/etc/azure/azure-registry-config.yaml"
      anonymous-access: "true"
```

Reply:
Spoke with @liggitt offline, we're going to put DockerConfig as is in the response so this shouldn't be a problem